kernel/src/hil/symmetric_encryption.rs

```diff
@@ -1,42 +1,87 @@
-//! 2. `set_key()`
-//! 3. `aes128_crypt_ctr()`: can be used arbitrary number of times.
+//! see boards/imix/src/aes_test.rs for example usage
 
 use returncode::ReturnCode;
 
-pub trait SymmetricEncryption {
-    /// Set the callback client.
-    fn set_client(&self, client: &'static Client);
+/// Implement this trait and use `set_client()` in order to receive callbacks from an `AES128`
+/// instance.
+pub trait Client {
+    fn crypt_done(&self);
+}
+
+/// The number of bytes used for AES block operations.  Keys and IVs must have this length,
+/// and encryption/decryption inputs must be have a multiple of this length.
+pub const AES128_BLOCK_SIZE: usize = 16;
+
+pub trait AES128<'a> {
+    /// Enable the AES hardware; must be called before any other methods
+    fn enable(&self);
+
+    /// Disable the AES hardware
+    fn disable(&self);
 
-    /// Initialization of the chip register
-    fn init(&self);
+    /// Set the client instance which will receive `crypt_done()` callbacks
+    fn set_client(&'a self, client: &'a Client);
 
-    /// Configure encryption/decryption key
-    /// assumes that key size is 16, 24 or 32 bytes
-    fn set_key(&self, key: &'static mut [u8], len: usize) -> &'static mut [u8];
+    /// Set the encryption key; returns `EINVAL` if length is not `AES128_BLOCK_SIZE`
+    fn set_key(&'a self, key: &'a [u8]) -> ReturnCode;
 
-    /// encryption and decryption for aes in counter mode
-    /// because only the encryption-mode of the cipher only one method is needed
-    /// other chips perhaps only ignore "init_ctr" and assume all is performed in HW
-    fn aes128_crypt_ctr(&self, data: &'static mut [u8], init_ctr: &'static mut [u8], len: usize);
+    /// Set the IV (or initial counter); returns `EINVAL` if length is not `AES128_BLOCK_SIZE`
+    fn set_iv(&'a self, iv: &'a [u8]) -> ReturnCode;
+
+    /// Set the source buffer.  If this is full, the encryption
+    /// input will be this entire buffer, and its size must match
+    /// `stop_index - start_index` when `crypt()` is called.
+    /// If this is empty, the destination buffer will be read
+    /// to provide the plaintext input.
+    fn set_source(&'a self, buf: Option<&'a [u8]>) -> ReturnCode;
+
+    /// Set the destination buffer.  If `set_source()` has not
+    /// been used to pass a source buffer, this buffer will also
+    /// provide the encryption input, which will be overwritten.
+    /// The option should be full whenever `crypt()` is called.
+    /// Returns SUCCESS if the buffer was installed, or EBUSY
+    /// if the encryption unit is still busy.
+    fn put_dest(&'a self, dest: Option<&'a mut [u8]>) -> ReturnCode;
+
+    /// Return the destination buffer, if any.
+    /// Returns EBUSY if the encryption unit is still busy.
+    fn take_dest(&'a self) -> Result<Option<&'a mut [u8]>, ReturnCode>;
+
+    /// Begin a new message (with the configured IV) when `crypt()` is next
+    /// called.  Multiple calls to `put_data()` and `crypt()` may be made
```

Comment:
I think `put_data` might refer to an older name for, perhaps, `set_source`?

Reply:
Oops, ok fixed